substrate/frame/delegated-staking/src/tests.rs

```diff
@@ -0,0 +1,647 @@
+				&agent,
+				&i,
+				slash,
+				Some(400)
+			));
+
+			// balance after slash.
+			assert_eq!(DelegatedStaking::held_balance_of(&i), unslashed_balance - slash);
+			// pending slash is reduced by the amount slashed.
+			assert_eq!(get_agent(&agent).ledger.pending_slash, initial_pending_slash - slash);
+			// reporter get 10% of the slash amount.
+			assert_eq!(
+				Balances::free_balance(reporter) - old_reporter_balance,
+				<Staking as StakingInterface>::slash_reward_fraction() * slash,
+			);
+			// update old balance
+			old_reporter_balance = Balances::free_balance(reporter);
+		}
+
+		// nothing to slash anymore
+		assert_eq!(get_agent(&agent).ledger.pending_slash, 0);
+
+		// cannot slash anymore
+		assert_noop!(
+			<DelegatedStaking as DelegationInterface>::delegator_slash(&agent, &350, 1, None),
+			Error::<T>::NothingToSlash
+		);
+	});
+}
+
+/// Integration tests with pallet-staking.
+mod staking_integration {
+	use super::*;
+	use pallet_staking::RewardDestination;
+	use sp_staking::Stake;
+
+	#[test]
+	fn bond() {
+		ExtBuilder::default().build_and_execute(|| {
+			let agent: AccountId = 99;
+			let reward_acc: AccountId = 100;
+			assert_eq!(Staking::status(&agent), Err(StakingError::<T>::NotStash.into()));
+
+			// set intention to become an agent
+			fund(&agent, 100);
+			assert_ok!(DelegatedStaking::register_agent(
+				RawOrigin::Signed(agent).into(),
+				reward_acc
+			));
+			assert_eq!(DelegatedStaking::stakeable_balance(&agent), 0);
+
+			let mut delegated_balance: Balance = 0;
+
+			// set some delegations
+			for delegator in 200..250 {
+				fund(&delegator, 200);
+				assert_ok!(DelegatedStaking::delegate_to_agent(
+					RawOrigin::Signed(delegator).into(),
+					agent,
+					100
+				));
+				delegated_balance += 100;
+				assert_eq!(
+					Balances::balance_on_hold(&HoldReason::StakingDelegation.into(), &delegator),
+					100
+				);
+				assert_eq!(DelegatedStaking::delegator_balance(&delegator), 100);
+
+				let agent_obj = get_agent(&agent);
+				assert_eq!(agent_obj.ledger.stakeable_balance(), delegated_balance);
+				assert_eq!(agent_obj.available_to_bond(), 0);
+				assert_eq!(agent_obj.bonded_stake(), delegated_balance);
+			}
+
+			assert_eq!(Staking::stake(&agent).unwrap(), Stake { total: 50 * 100, active: 50 * 100 })
+		});
+	}
+
+	#[test]
+	fn withdraw_test() {
+		ExtBuilder::default().build_and_execute(|| {
+			// initial era
+			start_era(1);
+			let agent: AccountId = 200;
+			let reward_acc: AccountId = 201;
+			let delegators: Vec<AccountId> = (301..=350).collect();
+			let total_staked =
+				setup_delegation_stake(agent, reward_acc, delegators.clone(), 10, 10);
+
+			// lets go to a new era
+			start_era(2);
+
+			assert!(eq_stake(agent, total_staked, total_staked));
+			// Withdrawing without unbonding would fail.
+			assert_noop!(
+				DelegatedStaking::release_delegation(RawOrigin::Signed(agent).into(), 301, 50, 0),
```

Comment:
?

Reply:
What is the question. :D
Don't like the error?